### PR TITLE
Add build and run tests in place

### PR DESCRIPTION
**Pull request by stcui007**
Add in place automated test and in framework automated workflow codes.

Please see automated tests here: https://github.com/stcui007/SoilFreezeThaw/pull/1

## Additions

.github/workflows/build_and_run_tests.yml
.github/workflows/ngen_integration.yaml
realizations/realization_multi_ngenCI.json

## Removals

-

## Changes

-

## Testing

Passes all automated tests. Please see here: https://github.com/stcui007/SoilFreezeThaw/pull/1

## Notes

-

## Todos

-

## Checklist

- [x ] PR has an informative and human-readable title
- [x ] Changes are limited to a single goal (no scope creep)
- [x ] Code can be automatically merged (no conflicts)
- [x ] Code follows project standards (link if applicable)
- [x ] Passes all existing automated tests
- [x ] Any _change_ in functionality is tested
- [x ] New functions are documented (with a description, list of inputs, and expected output)
- [x ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [x ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: .github/workflows/build_and_run_tests.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+      fail-fast: false
+    runs-on: ${{ matrix.os }}
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+    - name: Checkout the commit
+      uses: actions/checkout@v2
```
Comment (program--) [inline .github/workflows/build_and_run_tests.yml:33]: Should be `v4` not `v2`

**Comment by stcui007** [comment]
Yes. Thanks. I'll correct that.

**Comment by stcui007** [comment]
Have corrected the version number.

File: .github/workflows/build_and_run_tests.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      run: |
+        cd tests
+        if [ ${{ runner.os }} == 'macOS' ]
+        then
+          brew install gnu-sed
+          #PATH="$(brew --prefix)/opt/gnu-sed/libexec/gnubin:$PATH"
+          PATH="/usr/local/opt/gnu-sed/libexec/gnubin:$PATH"
+        fi
+        sed -i 's/\${CXX}/g++/' run_unittest.sh
+        ./run_unittest.sh
```
Comment (ajkhattak) [inline .github/workflows/build_and_run_tests.yml:69]: In practice, we always run unittests before running full model, so I would suggest move this up to run before `Build and Run Standalone`

File: .github/workflows/ngen_integration.yaml [hunk 3]
```diff
@@ -0,0 +1,132 @@
+# Test ngen-cfe integration
```
Comment (ajkhattak) [inline .github/workflows/ngen_integration.yaml:1]: replace `ngen-cfe` with `ngen-sft`

File: .github/workflows/ngen_integration.yaml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  # Obtained from https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources
+  # TODO: add something later that can check the runners for hyperthreading
+  LINUX_NUM_PROC_CORES: 2
+  MACOS_NUM_PROC_CORES: 3
+  ASAN_OPTIONS: detect_leaks=false
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  # Run general unit tests in linux environment
+  test_surfacebmi_plus_cfe:
```
Comment (ajkhattak) [inline .github/workflows/ngen_integration.yaml:23]: replace `test_surfacebmi_plus_cfe` with `test_sft` or something more relevant

File: .github/workflows/ngen_integration.yaml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        uses: actions/checkout@v4
+        with:
+          repository: noaa-owp/ngen
+
+      - name: Build Surfacebmi
+        id: submod_build_1
+        uses: ./.github/actions/ngen-submod-build
+        with:
+          mod-dir: " extern/noah-owp-modular/"
+          targets: "surfacebmi"
```
Comment (ajkhattak) [inline .github/workflows/ngen_integration.yaml:76]: I guess this is not needed as our existing tests don't include `noah-owp-modular`

File: .github/workflows/ngen_integration.yaml [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+      - name: Build ISO C Fortran BMI
+        id: submod_build_2
+        uses: ./.github/actions/ngen-submod-build
+        with:
+          mod-dir: "extern/iso_c_fortran_bmi/"
```
Comment (ajkhattak) [inline .github/workflows/ngen_integration.yaml:82]: not sure here. do we need this for nextgen build? if not, then probably this can also be removed.

**Review by ajkhattak — CHANGES_REQUESTED** [review]

File: .github/workflows/build_and_run_tests.yml (hunk 2 above)
Comment (stcui007) [reply .github/workflows/build_and_run_tests.yml:69]: Yeah, right. I probably followed the order in the INSTALL.md. Change made.

File: .github/workflows/ngen_integration.yaml (hunk 4 above)
Comment (stcui007) [reply .github/workflows/ngen_integration.yaml:23]: Done.

File: .github/workflows/ngen_integration.yaml (hunk 6 above)
Comment (stcui007) [reply .github/workflows/ngen_integration.yaml:82]: If there is no fortran code, this is probably not needed. Will test it.

**Review by ajkhattak — APPROVED** [review]

**Pull request merged**